advanced/Scripts/piholeDebug.sh

```diff
@@ -9,534 +9,1071 @@
+PIHOLE_WEB_SERVER_ERROR_LOG_FILE="${WEB_SERVER_LOG_DIRECTORY}/error.log"
+
+# An array of operating system "pretty names" that we officialy support
+# We can loop through the array at any time to see if it matches a value
+SUPPORTED_OS=("Raspbian" "Ubuntu" "Fedora" "Debian" "CentOS")
+
+# Store Pi-hole's processes in an array for easy use and parsing
+PIHOLE_PROCESSES=( "dnsmasq" "lighttpd" "pihole-FTL" )
+
+# Store the required directories in an array so it can be parsed through
+REQUIRED_DIRECTORIES=(${CORE_GIT_DIRECTORY}
+${CRON_D_DIRECTORY}
+${DNSMASQ_D_DIRECTORY}
+${PIHOLE_DIRECTORY}
+${PIHOLE_SCRIPTS_DIRECTORY}
+${BIN_DIRECTORY}
+${RUN_DIRECTORY}
+${LOG_DIRECTORY}
+${WEB_SERVER_LOG_DIRECTORY}
+${WEB_SERVER_CONFIG_DIRECTORY}
+${HTML_DIRECTORY}
+${WEB_GIT_DIRECTORY}
+${BLOCK_PAGE_DIRECTORY})
+
+# Store the required directories in an array so it can be parsed through
+mapfile -t array <<< "$var"
+REQUIRED_FILES=(${PIHOLE_CRON_FILE}
+${PIHOLE_DNS_CONFIG_FILE}
+${PIHOLE_DHCP_CONFIG_FILE}
+${PIHOLE_WILDCARD_CONFIG_FILE}
+${WEB_SERVER_CONFIG_FILE}
+${PIHOLE_DEFAULT_AD_LISTS}
+${PIHOLE_USER_DEFINED_AD_LISTS}
+${PIHOLE_BLACKLIST_FILE}
+${PIHOLE_BLOCKLIST_FILE}
+${PIHOLE_INSTALL_LOG_FILE}
+${PIHOLE_RAW_BLOCKLIST_FILES}
+${PIHOLE_LOCAL_HOSTS_FILE}
+${PIHOLE_LOGROTATE_FILE}
+${PIHOLE_SETUP_VARS_FILE}
+${PIHOLE_WHITELIST_FILE}
+${PIHOLE_COMMAND}
+${PIHOLE_COLTABLE_FILE}
+${FTL_PID}
+${FTL_PORT}
+${PIHOLE_LOG}
+${PIHOLE_LOG_GZIPS}
+${PIHOLE_DEBUG_LOG}
+${PIHOLE_FTL_LOG}
+${PIHOLE_WEB_SERVER_ACCESS_LOG_FILE}
+${PIHOLE_WEB_SERVER_ERROR_LOG_FILE})
+
+source_setup_variables() {
+  # Display the current test that is running
+  log_write "\n${COL_LIGHT_PURPLE}*** [ INITIALIZING ]${COL_NC} Sourcing setup variables"
+  # If the variable file exists,
+  if ls "${PIHOLE_SETUP_VARS_FILE}" 1> /dev/null 2>&1; then
+    log_write "${INFO} Sourcing ${PIHOLE_SETUP_VARS_FILE}...";
+    # source it
+    source ${PIHOLE_SETUP_VARS_FILE}
+  else
+    # If it can't, show an error
+    log_write "${PIHOLE_SETUP_VARS_FILE} ${COL_LIGHT_RED}does not exist or cannot be read.${COL_NC}"
+  fi
 }
 
-header_write() {
-  log_echo ""
-  log_echo "---= ${1}"
-  log_write ""
+make_temporary_log() {
+  # Create a random temporary file for the log
+  TEMPLOG=$(mktemp /tmp/pihole_temp.XXXXXX)
+  # Open handle 3 for templog
+  # https://stackoverflow.com/questions/18460186/writing-outputs-to-log-file-and-console
+  exec 3>"$TEMPLOG"
+  # Delete templog, but allow for addressing via file handle
+  # This lets us write to the log without having a temporary file on the drive
+  rm "$TEMPLOG"
 }
 
-file_parse() {
-    while read -r line; do
-		  if [ ! -z "${line}" ]; then
-			  [[ "${line}" =~ ^#.*$  || ! "${line}" || "${line}" == "WEBPASSWORD="* ]] && continue
-				log_write "${line}"
-			fi
-		done < "${1}"
-		log_write ""
+log_write() {
+  # echo arguments to both the log and the console
+  echo -e "${@}" | tee -a /proc/$$/fd/3
```

Comment:
Stripping the character codes off in the `log_write` function will stop the colours from being visible, so I'll implement the `sed` command somewhere else.  This should also help not having to make a new subshell everytime the `log_write` is called.  Should be easier to just strip the characters off once before upload.